src/CityOnPlanet.cpp

```diff
@@ -202,123 +312,345 @@ CityOnPlanet::~CityOnPlanet()
-	} else {
-		population *= 100;
-		cityradius = (population < 250) ? 250 : ((population > START_SEG_SIZE_NO_ATMO) ? START_SEG_SIZE_NO_ATMO : population);
-	}
-
-	citybuildinglist_t *buildings = &s_buildingList;
-	vector3d cent = p;
-	const int cellsize_i = 80;
-	const double cellsize = double(cellsize_i);						// current widest building = 92
-	const double bodyradius = planet->GetSystemBody()->GetRadius(); // cache for bodyradius value
-
-	static const int gmid = (cityradius / cellsize_i);
-	static const int gsize = gmid * 2;
-
-	assert((START_SEG_SIZE / cellsize_i) < 100);
-	assert((START_SEG_SIZE_NO_ATMO / cellsize_i) < 100);
-	uint8_t cellgrid[200][200];
-	std::memset(cellgrid, 0, sizeof(cellgrid));
-
-	// calculate the size of the station model
-	const int x1 = floor(aabb.min.x / cellsize);
-	const int x2 = ceil(aabb.max.x / cellsize);
-	const int z1 = floor(aabb.min.z / cellsize);
-	const int z2 = ceil(aabb.max.z / cellsize);
-
-	// Clear the cells where the station is
-	for (int x = 0; x <= gsize; x++) {
-		for (int z = 0; z <= gsize; z++) {
-			const int zz = z - gmid;
-			const int xx = x - gmid;
-			if (zz > z1 && zz < z2 && xx > x1 && xx < x2)
-				cellgrid[x][z] = 1;
-		}
+	AddStaticGeomsToCollisionSpace();
+}
+
+void CityOnPlanet::Generate(SpaceStation *station)
+{
+	PROFILE_SCOPED()
+	Profiler::Clock _genTimer;
+	_genTimer.Start();
+
+	// Create the acceleration grid structure
+	// ==========================================
+
+	// Retrieve the station parameters
+	uint8_t stationSize[2] = {};
+	GetModelSize(station->GetAabb(), stationSize);
+
+	uint32_t stationSizeMax = std::max<uint32_t>(stationSize[0] / 2, stationSize[1] / 2);
+
+	// update the allowed radius to ensure the spaceport isn't larger than the entire city
+	m_cityRadius += stationSizeMax * CELLSIZE;
+
+	// For now, just assume the city is a circle around the spaceport in the center
+	const uint32_t cityExtents = (m_cityRadius / CELLSIZE);
+	m_citySize = cityExtents * 2;
+	m_gridPitch = std::ceil(m_citySize / 8.0);
+
+	Log::Verbose("Generating City for spacestation {}", station->GetSystemBody()->GetName());
+	Log::Verbose("\tpopulation: {0} size {1}x{1}c (radius {2})",
+		m_body->GetPopulation(), m_citySize, m_cityRadius);
+
+
+	// ensure divisible by 8 bytes for fast 64-bit lookup
+	if (m_gridPitch & 7) {
+		m_gridPitch = (m_gridPitch + 7) & ~7;
 	}
+	m_gridLen = m_gridPitch * m_citySize;
+
+	m_gridBitset.reset(new uint8_t[m_gridLen]);
+	std::memset(m_gridBitset.get(), 0, m_gridLen);
+
+	// Setup the station model position relative to the city grid
+	// ==========================================
+
+	// get the increment vectors to multiply X/Z grid coordinates by
+	vector3d incX = station->GetOrient().VectorX() * CELLSIZE;
+	vector3d incZ = station->GetOrient().VectorZ() * CELLSIZE;
+
+	// top-left corner of the grid is -X, -Z relative to station position at center
+	m_gridOrigin = station->GetPosition() - incX * cityExtents - incZ * cityExtents;
```

Comment:
On pure intuition, I rewrote it like this:
`m_gridOrigin = station->GetPosition() - incX * (cityExtents + 0.5) - incZ * (cityExtents + 0.5);`
before
![image](https://user-images.githubusercontent.com/18342621/204900979-9972418b-75cc-4682-ba89-9b7bd2c84c8d.png)

after
![image](https://user-images.githubusercontent.com/18342621/204900630-cef4c1b7-21b2-4113-8d8b-e1af736be78f.png)

Reply:
Good catch, thank you! That subtle offset always bugged me.